ocs_ci/ocs/resources/namespacestore.py

```diff
@@ -0,0 +1,155 @@
+log = logging.getLogger(__name__)
+
+
+class NamespaceStore:
+    """
+    A class that represents NamespaceStore objects
+
+    """
+
+    def __init__(
+        self,
+        name,
+        method,
+        uls_name=None,
+        secret_name=None,
+        mcg_obj=None,
+    ):
+        self.name = name
+        self.method = method
+        self.uls_name = uls_name
+        self.secret_name = secret_name
+        self.mcg_obj = mcg_obj
+
+    def delete(self):
+        log.info(f"Cleaning up namespacestore {self.name}")
+
+        if self.method == "oc":
+            OCP(
+                kind="namespacestore", namespace=config.ENV_DATA["cluster_namespace"]
+            ).delete(resource_name=self.name)
+        elif self.method == "cli":
+
+            def _cli_deletion_flow():
+                try:
+                    self.mcg_obj.exec_mcg_cmd(f"namespacestore delete {self.name}")
+                    return True
+                except CommandFailed as e:
+                    if "being used by one or more buckets" in str(e).lower():
+                        log.warning(
+                            f"Deletion of {self.name} failed because it's being used by a bucket. "
+                            "Retrying..."
+                        )
+                        return False
+
+            sample = TimeoutSampler(
+                timeout=120,
+                sleep=20,
+                func=_cli_deletion_flow,
+            )
+            if not sample.wait_for_func_status(result=True):
+                log.error(f"Failed to {self.name}")
+                raise TimeoutExpiredError
+
+        log.info(f"Verifying whether namespacestore {self.name} exists after deletion")
+        ns_deleted_successfully = False
+
+        try:
+            if self.method == "oc":
+                OCP(
+                    kind="namespacestore",
+                    namespace=config.ENV_DATA["cluster_namespace"],
+                    resource_name=self.name,
+                ).get()
+            elif self.method == "cli":
+                self.mcg_obj.exec_mcg_cmd(f"namespacestore status {self.name}")
+
+        except CommandFailed as e:
+            if "Not Found" in str(e) or "NotFound" in str(e):
+                ns_deleted_successfully = True
+            else:
+                raise
+
+        assert (
+            ns_deleted_successfully
+        ), f"Namespacestore {self.name} was not deleted successfully"
+
+
+def namespace_store_factory(request, cld_mgr, mcg_obj, cloud_uls_factory):
+    """
+    Create a namespace store factory. Calling this fixture creates a new namespace store.
+
+    """
+    created_nss = []
+
+    cmdMap = {"oc": mcg_obj.create_namespace_store, "cli": ""}  # TODO
+
+    endpointMap = {
+        constants.AWS_PLATFORM: constants.MCG_NS_AWS_ENDPOINT,
+        constants.AZURE_PLATFORM: constants.MCG_NS_AZURE_ENDPOINT,
+        constants.RGW_PLATFORM: RGW().get_credentials()[0],
+    }
+
+    def _create_nss(method, nss_dict):
```

Comment:
Please add docstring describing parameters used by this factory function and format of nss_dict.
I would prefer if information about store were not provided in dictionary but in parameters so it's clear and only one store is created upon one call of factory function.

Reply:
>Please add docstring describing parameters used by this factory function and format of nss_dict.

Done

>I would prefer if information about store were not provided in dictionary but in parameters so it's clear and only one store is created upon one call of factory function.`

That's gonna be heavy - so feel free to reach out for a call if you feel it's needed -
I find this problematic for two reasons:
1. Personally, I highly doubt the replacement of the dicts with variables will make things more readable. Actually, the opposite is the same. When parametrizing variables, the parameters seem meaningless and disconnected from each other, and it's not clear which is which.
i.e. - `"OC", "Single", "aws", "eu-central-1"` 
seems unclearer than
```
"interface": "OC",
  "namespace_policy_dict": {
      "type": "Single",
      "namespacestore_dict": {
          "aws": [(1, "eu-central-1")]
```
2. This being the main reason IMO - such a change will lead to _substantial_ code duplication and complexity.
It means that each and every test that uses more than two backing/NSstores will have to include, at the test level, an iterator that checks how many stores need to be created, just to call the factory function on them one by one - solely for the sake of not using a dictionary. On top of that, the tests will no longer be able to remain simple one-liners. The stores would have to be created manually, then a bucketclass containing them all will need to be created manually, and *then* passed to the bucket factory. This will be done _for every_ test that requires more than a single backingstore or namespace store - adding complexity, duplication, length, and IMO confusion.

Reply:
1. This can be solved by using named parameters. e.g. `store = namespace_store_factory(cloud_name='asw', region='eu-central-1', type='Single', interface='OC')`
2. I think that those iterators in tests would make it more explicit. I would also prefer if everything was not done in one step so that we have more freedom in creating negative test cases and can access and update used namespace stores and bucketclasses in tests explicitly before bucket_factory is called. That said, it can be changed in another PR if needed.

Reply:
Is there any pending item to be done as part of this PR?
If so, suggest opening an issue and taking it from there